### ISSUE TITLE
"openssl:Error: 'sha512' is an invalid command" on Mac OS X 10.6.8 (Snow Leopard)

### ISSUE THREAD

**Issue by maroneze** [ocaml/opam#3930]
On macOS 10.6.8, `openssl` does not know the `sha512` command, but it does accept `openssl dgst -sha512`.

While trying to migrate from 2.0.0 to 2.0.4, I have several `openssl error`s, for instance `opam-state.2.0.4`, `opam-solver.2.0.4`, `opam-core.2.0.4`, etc. All of them due to the issue with SHA512.

It seems to me, naively, after some quick tests, that replacing `openssl sha512` with `openssl dgst -sha512` should work in all platforms in which it does already, plus macOS 10.6.8 (at least).

But since it's a very old system, if you decide it's not worth doing, no problems. I'm not even sure I'll be able to upgrade, since most things in Homebrew are broken in this old macOS.

```
# opam config report
# opam-version      2.0.0 
# self-upgrade      no
# system            arch=x86_32 os=macos os-distribution=homebrew os-version=10.6.8
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              4
# repositories      1 (http) (default repo at 4c2bd0f9)
# pinned            0
# current-switch    4.02.3
```

**Comment by AltGr** [comment]
Thanks for reporting!

Note that opam also has its own implementation of sha512 (for reasons of compat, and faster hash of small data chunks). Unfortunately, it just checks for the _existence_ of `openssl` — we weren't aware of versions that didn't support `sha512` as a direct subcommand.

But there is an environment variable `OPAMUSEOPENSSL` that could help here: setting it to `0` should be a viable workaround.

Anyway, using `openssl dgst` instead seems like a good idea for compat: thanks.

**Comment by AltGr** [comment]
Oh, I was just told this was already fixed in 2.0.1. Thanks anyway :)

**Comment by rjbou** [comment]
I you upgrade to opam 2.0.1, you'll have the fix (#3543): opam roll back to its own implementation in case of error. But this works only for opam itself, not for the install script (cf. PR #3516).
Related also to https://github.com/ocaml/opam-repository/pull/12613.

We'll look about the portability of `dgst` and integrate it then.

**Comment by maroneze** [comment]
Sorry, I was going to check with a newer opam, but my Snow Leopard is just dying of old age (certificate issues, Homebrew unsupported, etc), so after hours trying to fix it, I finally got someone to upgrade it to El Capitan. Good to know it's already fixed!
